frontend/src/mixins/autocomplete.ts

```diff
@@ -0,0 +1,222 @@
+import {LitElement} from 'lit';
+import {Constructor} from './utils';
+import {property} from 'lit/decorators.js';
+import {DebounceMixin, DebounceMixinInterface} from './debounce';
+
+/**
+ * Interface for the Autocomplete mixin.
+ */
+export interface AutocompleteMixinInterface extends DebounceMixinInterface {
+  inputName: string;
+  options: AutocompleteOption[];
+  value: string;
+  currentIndex: number;
+  getOptionIndex: number;
+  visible: boolean;
+  isLoading: boolean;
+  currentOption: AutocompleteOption | undefined;
+
+  onInput(event: InputEvent): void;
+  handleInput(value: string): void;
+  blurInput(): void;
+  resetInput(): void;
+  submit(isSuggestion?: boolean): void;
+  getAutocompleteValueByIndex(index: number): string;
+  handleArrowKey(direction: 'up' | 'down'): void;
+  handleEnter(event: KeyboardEvent): void;
+  handleEscape(): void;
+  onKeyDown(event: KeyboardEvent): void;
+  onClick(index: number): () => void;
+  onFocus(): void;
+  onBlur(): void;
+}
+/**
+ * Type for autocomplete option.
+ */
+export type AutocompleteOption = {
+  value: string;
+  label: string;
+};
+/**
+ * Type for autocomplete result.
+ */
+export type AutocompleteResult = {
+  value: string;
+  label?: string;
+};
+
+export const AutocompleteMixin = <T extends Constructor<LitElement>>(
+  superClass: T
+): Constructor<AutocompleteMixinInterface> & T => {
+  class AutocompleteMixinClass extends DebounceMixin(superClass) {
+    @property({attribute: 'input-name'})
+    inputName = 'autocomplete';
+
+    /**
+     * The options for the autocomplete.
+     * It is provided by the parent component.
+     */
+    @property({attribute: false, type: Array})
+    options: AutocompleteOption[] = [];
+
+    // selected values
+    @property()
+    value = '';
+
+    @property({attribute: false})
+    currentIndex = 0;
+
+    @property({attribute: false})
+    visible = false;
+
+    @property({attribute: false})
+    isLoading = false;
+
+    /**
+     * This method is used to get the current index.
+     * It remove the offset of 1 because the currentIndex is 1-based.
+     * @returns {number} The current index.
+     */
+    get getOptionIndex() {
+      return this.currentIndex - 1;
+    }
+
+    get currentOption() {
+      return this.options[this.getOptionIndex];
+    }
+
+    /**
+     * Handles the input event on the autocomplete and dispatch custom event : "autocomplete-input".
+     * @param {InputEvent} event - The input event.
+     */
+    onInput(event: InputEvent) {
+      const value = (event.target as HTMLInputElement).value;
+      this.value = value;
+      this.handleInput(value);
+    }
+
+    handleInput(value: string) {
+      throw new Error(`handleInput method must be implemented with ${value}`);
```

Comment:
Just to be helpful to the developer, identify the class :-)

```suggestion
      throw new Error(`handleInput method must be implemented for ${this} with ${value}`);
```